pkg/utils/metadata_test.go

```diff
@@ -67,6 +70,66 @@ func TestProjectID(t *testing.T) {
 	}
 }
 
+func TestProjectIDOrDefault(t *testing.T) {
+	testCases := map[string]struct {
+		want  string
+		data  testingMetadataClient.TestClientData
+		input string
+		env   string
+		error bool
+	}{
+		"project id exists": {
+			want:  "testing-project",
+			data:  testingMetadataClient.TestClientData{},
+			input: "testing-project",
+			error: false,
+		},
+		"Successfully get project id from env": {
+			want:  "testing-project",
+			data:  testingMetadataClient.TestClientData{},
+			env:   "testing-project",
+			input: "",
+			error: false,
+		},
+		"Successfully get project id from client": {
+			want:  testingMetadataClient.FakeProjectID,
+			data:  testingMetadataClient.TestClientData{},
+			input: "",
+			error: false,
+		},
+		"project id doesn't exist, get project id failed": {
+			want: "",
+			data: testingMetadataClient.TestClientData{
+				ProjectIDErr: fmt.Errorf("error getting project id"),
+			},
+			input: "",
+			error: true,
+		},
+	}
+
+	for n, tc := range testCases {
+		t.Run(n, func(t *testing.T) {
+			// Save and restore changed variables
+			origEnv := os.Getenv(ProjectIDEnvKey)
```

Comment:
Instead of changing and then restoring the `ProjectIDEnvKey`, is it possible that we write a fake os so that we can avoid changing the env variables in UT?

Reply:
I like this idea. Do you know a way to do it? Let me dig into how env is handled in go UT. I think it is safe to always clear env variables in UT as there should not be such assumption anyway. See how `evnconfig` package test it: https://github.com/kelseyhightower/envconfig/blob/c974cae29cf5b543b30f7ba91f8ce9649cd69818/envconfig_test.go#L91

Reply:
If it's safe to change env variables in UT, this comment will just be a nit.
IMO, based on currently implementation, it will be easier if you can write a fake envConfig by yourself.

Reply:
This code is removed, we now only infer project id once during startup.